pkg/datasource/coinmarketcap/v1/client.go

```diff
@@ -0,0 +1,111 @@
+package v1
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/c9s/requestgen"
+)
+
+const baseURL = "https://pro-api.coinmarketcap.com"
+
+type RestClient struct {
+	BaseURL *url.URL
+	Client  *http.Client
+
+	apiKey string
+}
+
+func New() *RestClient {
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		panic(err)
+	}
+
+	return &RestClient{
+		BaseURL: u,
+		Client:  &http.Client{},
+	}
+}
+
+func (c *RestClient) Auth(apiKey string) {
+	c.apiKey = apiKey
+}
+
+func (c *RestClient) NewRequest(ctx context.Context, method string, refURL string, params url.Values, payload interface{}) (*http.Request, error) {
+	rel, err := url.Parse(refURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if params != nil {
+		rel.RawQuery = params.Encode()
+	}
+
+	pathURL := c.BaseURL.ResolveReference(rel)
+
+	body, err := castPayload(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println(pathURL.String())
+
+	return http.NewRequestWithContext(ctx, method, pathURL.String(), bytes.NewReader(body))
+}
+
+func (c *RestClient) SendRequest(req *http.Request) (*requestgen.Response, error) {
```

Comment:
check requestgen.BaseApiClient, it implements the SendRequest method and the NewRequest method. ;)

Reply:
done